### DISCUSSION TITLE
For strongSwan, how to re-establish the IPsec tunnel automatically always from one end after it breaks due to PPPoE IP address changes?

### DISCUSSION THREAD

**Discussion by TomHsiung** [strongswan/strongswan#1695] [Q&A]
I know `dpdaction=restart` enable the re-establishment function. However, this parameter is not a local option and both ends need to agree on it. My goal is that the re-establishment is always initiated from one end. Can we achieve this goal?

**Comment by tobiasbrunner** [comment]
> However, this parameter is not a local option and both ends need to agree on it.

No, it's **only** a local option and the peers **don't** have to agree on it at all.

> Can we achieve this goal?

Configure it only on one end (use `dpdaction=clear` on the other). But be aware that plaintext traffic might leak without established SA.

Also, why can't the SA be updated via MOBIKE after an address change?

**Reply by TomHsiung** [reply]
> why can't the SA be updated via MOBIKE after an address change?

I think it is due to the fact that I use domain names instead of IP address as the tunnel ids. After PPPoE breaks up and reconnects, there is a delay in update the new IP address to DNS records service. And I think this delay might cause MOBIKE not working.

PS: for site-to-site mode, both ends use domain names as their id in `ipsec.conf`.

**Reply by tobiasbrunner** [reply]
> I think it is due to the fact that I use domain names instead of IP address as the tunnel ids.

How is that relevant in regards to MOBIKE? There is no mapping from identities to IPs and the whole point of MOBIKE is that the existing IKE_SA is updated (i.e. there is no reauthentication where identities would be relevant).

> After PPPoE breaks up and reconnects, there is a delay in update the new IP address to DNS records service.

So the IP of the server changed too? Or on which end is the PPPoE connection?

> And I think this delay might cause MOBIKE not working.

Read the logs to see what's going on.

> PS: for site-to-site mode, both ends use domain names as their id in `ipsec.conf`.

Again, the identities don't matter in regards to MOBIKE, they are used as is.

**Reply by TomHsiung** [reply]
> So the IP of the server changed too? Or on which end is the PPPoE connection?

The remote end has a static IP. The local end use PPPoE so its IP address is dynamic. I don't know. Maybe the re-establishment of PPPoE takes too long time longer than MOBIKE could tolerate. Or maybe I have not turned on MOBIKE function. But in the `ipsec.conf` file I did not configure `MOBIKE` parameter so it is in its default value.

Or the PPPoE link is just broken (off) when the IP changes, as if it was unplugged so the tunnel just broke.

**Reply by tobiasbrunner** [reply]
As I said before, read the logs to see what happens on the client (use the log levels given in #196 for more details).

**Reply by Kentzo** [reply]
If the PPPoE host is an initiator then MOBIKE should just work. Doesn't require any additional network adjustments. If it's a responder then it won't work: no way to communicate new IP to the initiator.

**Reply by tobiasbrunner** [reply]
@Kentzo That's not entirely true. The server may switch to a different endpoint via an update of the additional addresses set ([RFC 4555, section 3.6](https://datatracker.ietf.org/doc/html/rfc4555#section-3.6)), which could then trigger an address update from the client. However, the client might be behind a NAT or firewall that doesn't let those messages through from a new IP address.

**Reply by TomHsiung** [reply]
It is in site-to-site mode but the PPPoE host indeed initiates the IPsec tunnel.

**Reply by Kentzo** [reply]
@tobiasbrunner afaik current address must still be available for the responder to notify the initiator about the additional ones, which is not always possible

@TomHsiung on the pppoe machine, immediately after you disconnect, what is the status of SAs? Perhaps some event forces strongswan to flush them?

**Reply by tobiasbrunner** [reply]
> afaik current address must still be available for the responder to notify the initiator about the additional ones

Other than due to the potential issues with NATs/firewalls, no (read the part that starts with "There is one additional complication").

> Perhaps some event forces strongswan to flush them?

There is no code in strongSwan that flushes SAs, routes or routing rules.

According to Tom's latest post in the other thread, it seems more like something triggers a restart of the complete networking stack that causes the rules to get flushed. For instance, systemd-networkd, see https://github.com/systemd/systemd/issues/19106 - there is apparently a config option for it that could be disabled to avoid this (`ManageForeignRoutingPolicyRules`). Theoretically, we could listen for events regarding the removal of our rule and reinstall it, but not sure if events are triggered if rules are flushed.

**Reply by TomHsiung** [reply]
`@TomHsiung on the pppoe machine, immediately after you disconnect, what is the status of SAs? Perhaps some event forces strongswan to flush them?`

I don't know. However, for roadwarrior  mode, the interface does not lose the IP address, it just switches from one IP address to another (e.g., 5G to home WiFi). In contrast, the machine which initiates PPPoE might have lost its IP address during a short period of time when PPPoE status was update by the ISP. So during this time the PPPoE interface might has no IP address, which causes the MOBIKE to fail.

**Comment by TomHsiung** [comment]
route table 220 before unplugging
> default via R.R.R.R(remote VPN gateway) dev ppp0 proto static src 10.0.0.1 
> throw 10.0.0.0/24 proto static 
> throw 10.0.11.0/24 proto static 

route table 220 after re-plugging
> default via R.R.R.R(remote VPN gateway) dev ppp0 proto static src 10.0.0.1 
> throw 10.0.0.0/24 proto static

`ip route get x.x.x.x` before unplugging
> x.x.x.x via R.R.R.R dev ppp0 table 220 src 10.0.0.1 uid 1000 
>     cache

`ip route get x.x.x.x` after re-plugging
> x.x.x.x dev ppp0 src L.L.L.L(local VPN gateway public IP) uid 1000 
>     cache

I will add the log of this trial later.

**Reply by tobiasbrunner** [reply]
What is `10.0.11.0/24`? Is it relevant that its throw route is missing? Do you have configured bypass policies manually or via _bypass-lan_ plugin?

What does the main routing table look like before and after? Is there a change in `ip rule` before and after? What does `ip route get fibmatch x.x.x.x` return before and after?

**Reply by TomHsiung** [reply]
> What is 10.0.11.0/24? Is it relevant that its throw route is missing? Do you have configured bypass policies manually or via bypass-lan plugin?
`10.0.11.0/24` is a subnet. Between the strongSwan server (`10.0.0.1`) and the subnet there is a router.

I both wrote bypass policies manually and enabled the bypass plugin.
> # passthrough policy 1
> conn passthrough1
>         left=127.0.0.1
>         leftsubnet=10.0.0.0/24,10.0.11.0/24
>         rightsubnet=10.0.11.0/24,10.0.0.0/24
>         type=passthrough
>         authby=never
>         auto=route

And

> Shunted Connections:
> Bypass LAN 10.0.0.0/24:  10.0.0.0/24 === 10.0.0.0/24 PASS
> Bypass LAN ::1/128:  ::1/128 === ::1/128 PASS
> Bypass LAN fe80::/64:  fe80::/64 === fe80::/64 PASS

**Reply by tobiasbrunner** [reply]
> I both wrote bypass policies manually and enabled the bypass plugin.

Hm, why both? I'd recommend only using either the plugin or manual policies. However, I don't see `10.0.11.0/24` in the _bypass-lan_ policies, is that not a locally connected subnet? Also, since the _bypass-lan_ policies are very specific (local/remote is the same local subnet) you might need manual policies if you want to bypass traffic from one to some other subnet.

**Reply by TomHsiung** [reply]
> What does the main routing table look like before and after?

route table before

> default dev ppp0 scope link 
> default via 10.0.0.1 dev enp1s0 proto static 
> 10.0.0.0/24 dev enp1s0 proto kernel scope link src 10.0.0.1 
> 10.0.11.0/24 via 10.0.0.254 dev enp1s0 proto static 
> x.x.x.1 dev ppp0 proto kernel scope link src x.x.x.x

> Is there a change in ip rule before and after?

ip rule before

> 0:	from all lookup local
> 220:	from all lookup 220
> 32766:	from all lookup main
> 32767:	from all lookup default

> What does ip route get fibmatch x.x.x.x return before and after?

`ip route get fibmatch 10.5.96.4`

> default via remote.ipsec.gateway dev ppp0 table 220 proto static src 10.0.0.1

To be continued ...

**Reply by TomHsiung** [reply]
route table after

> default dev ppp0 scope link 
> default via 10.0.0.1 dev enp1s0 proto static 
> 10.0.0.0/24 dev enp1s0 proto kernel scope link src 10.0.0.1 
> 10.0.11.0/24 via 10.0.0.254 dev enp1s0 proto static 
> x.x.x.1 dev ppp0 proto kernel scope link src x.x.x.x 

ip rule after

> 0:	from all lookup local
> 32766:	from all lookup main
> 32767:	from all lookup default

ip route get fibmatch 10.5.96.5

> default dev ppp0 scope link

**Reply by TomHsiung** [reply]
Oh, after the unplugging and re-plugging, it is clear that the 220 disappeared from the ip rule list. I think it is the cause.

**Reply by tobiasbrunner** [reply]
Hm, that's weird. What exactly flushes those rules (and when)?

**Reply by TomHsiung** [reply]
IDK. Maybe you can make a try. Ubuntu 22.04 TLS.
